client/app/components/dashboards/Widget.jsx

```diff
@@ -0,0 +1,336 @@
+            onClick={() => this.refreshWidget(1)}
+            data-test="RefreshButton"
+          >
+            <i className={classNames('zmdi zmdi-refresh', { 'zmdi-hc-spin': refreshClickButtonId === 1 })} />{' '}
+            <TimeAgo date={updatedAt} />
+          </a>
+        )}
+        <span className="visible-print">
+          <i className="zmdi zmdi-time-restore" />{' '}{formatDateTime(updatedAt)}
+        </span>
+        {isPublic ? (
+          <span className="small hidden-print">
+            <i className="zmdi zmdi-time-restore" />{' '}<TimeAgo date={updatedAt} />
+          </span>
+        ) : (
+          <a
+            className="btn btn-sm btn-default pull-right hidden-print btn-transparent btn__refresh"
+            onClick={() => this.refreshWidget(2)}
+          >
+            <i className={classNames('zmdi zmdi-refresh', { 'zmdi-hc-spin': refreshClickButtonId === 2 })} />
+          </a>
+        )}
+        <a
+          className="btn btn-sm btn-default pull-right hidden-print btn-transparent btn__refresh"
+          onClick={this.expandWidget}
+        >
+          <i className="zmdi zmdi-fullscreen" />
+        </a>
+      </div>
+    );
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  renderWidgetVisualization() {
+    const { widget, filters } = this.props;
+    const widgetQueryResult = widget.getQueryResult();
+    const widgetStatus = widgetQueryResult && widgetQueryResult.getStatus();
+    switch (widgetStatus) {
+      case 'failed':
+        return (
+          <div className="body-row-auto scrollbox">
+            {widgetQueryResult.getError() && (
+              <div className="alert alert-danger m-5">
+                Error running query: <strong>{widgetQueryResult.getError()}</strong>
+              </div>
+            )}
+          </div>
+        );
+      case 'done':
+        return (
+          <div ng-switch-when="done" className="body-row-auto scrollbox">
+            <VisualizationRenderer
+              visualization={widget.visualization}
+              queryResult={widgetQueryResult}
+              filters={filters}
+            />
+          </div>
+        );
+      default:
+        return (
+          <div className="body-row-auto spinner-container">
+            <div className="spinner">
+              <i className="zmdi zmdi-refresh zmdi-hc-spin zmdi-hc-5x" />
+            </div>
+          </div>
+        );
+    }
+  }
+
+  renderWidget() {
+    const { widget } = this.props;
+    const widgetQueryResult = widget.getQueryResult();
+    const isRefreshing = widget.loading && !!(widgetQueryResult && widgetQueryResult.getStatus());
+
+    return (
+      <div className="tile body-container widget-visualization visualization" data-refreshing={isRefreshing}>
+        {this.renderWidgetHeader()}
+        {this.renderWidgetVisualization()}
+        {this.renderWidgetBottom()}
+      </div>
+    );
+  }
+
+  renderTextbox() {
```

Comment:
@gabrieldutra I've always felt there's an unnatural relationship between widget and textbox. A widget is a dashboard item but also a visualization widget; a textbox is a widget but only in the dashboard item aspect. Wdyt of a 1 level abstraction?
```
Widget (dashboard item, resize, reorder, menu button, remove button)
  |__ Visualization (title, viz, footer, refreshIndicator, autoHeight)
  |__ Textbox (markdown)
```

Reply:
My next commit is exactly something like that :grin:, I started yesterday, but didn't have much time.